### PR TITLE
fix: deploy smart pay

### PR DESCRIPTION
**Pull request by nuanyang233**
## Description

Closes #MF-3259 MF-3258 MF-3235 MF-3137 MF-3143 MF-3146 MF-3264 MF-3266 MF-3268 MF-3243

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx [hunk 1]
```diff
@@ -160,6 +168,21 @@ export default function ChangeOwner() {
...
+    const onCreatePersona = useCallback(() => {
+        browser.tabs.create({
+            active: true,
+            url: browser.runtime.getURL(`/dashboard.html#${DashboardRoutes.SignUp}`),
+        })
+        if (process.env.engine === 'firefox') {
+            window.close()
+        }
+        Services.Helper.removePopupWindow()
+    }, [])
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ChangeOwner/index.tsx]:
```suggestion
    const onCreatePersona = useCallback(async () => {
        browser.tabs.create({
            active: true,
            url: browser.runtime.getURL(`/dashboard.html#${DashboardRoutes.SignUp}`),
        })
        if (process.env.engine === 'firefox') {
            window.close()
        }
        await Services.Helper.removePopupWindow()
    }, [])
```

File: packages/mask/src/extension/popups/pages/Wallet/ContractInteraction/index.tsx [hunk 2]
```diff
@@ -350,7 +350,7 @@ const ContractInteraction = memo(() => {
 
         if (!gasCurrency || isNativeTokenAddress(gasCurrency)) return result
         if (!currencyRatio) return ZERO
-        return result.dividedBy(currencyRatio)
+        return new BigNumber(result.multipliedBy(currencyRatio).toFixed(0))
```
Comment (guanbinrui) [inline packages/mask/src/extension/popups/pages/Wallet/ContractInteraction/index.tsx]:
```suggestion
        return toFixed(multipliedBy(currencyRatio))
```

File: packages/plugins/EVM/src/state/TransactionFormatter/descriptors/ERC20.ts [hunk 3]
```diff
@@ -42,6 +42,7 @@ export class ERC20Descriptor implements TransactionDescriptor {
                         }
                     }
 
+                    console.log(parameters)
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/TransactionFormatter/descriptors/ERC20.ts]:
```suggestion
```

File: packages/plugins/SmartPay/src/hooks/useDeploy.tsx [hunk 4]
```diff
@@ -43,6 +42,7 @@ export function useDeploy(
     )
     const connection = useWeb3Connection(NetworkPluginID.PLUGIN_EVM, {
         providerType: ProviderType.MaskWallet,
+        account: undefined,
```
Comment (guanbinrui) [inline packages/plugins/SmartPay/src/hooks/useDeploy.tsx:45]: What's the difference?

File: packages/plugins/SmartPay/src/hooks/useDeploy.tsx [hunk 5]
```diff
@@ -101,29 +101,33 @@ export function useDeploy(
...
-                        updatedAt: new Date(),
-                    }).then(() => {
-                        onSuccess?.()
-                    })
-                } catch (error) {
-                    console.log(error)
-                }
+            return new Promise((resolve, reject) => {
+                return TransactionWatcher?.emitter.on('progress', async (_, txHash, status) => {
+                    try {
```
Comment (guanbinrui) [inline packages/plugins/SmartPay/src/hooks/useDeploy.tsx]: Redundant try-catch.

File: packages/plugins/SmartPay/src/hooks/useDeploy.tsx [hunk 6]
```diff
@@ -101,29 +101,33 @@ export function useDeploy(
...
+                        if (txHash !== hash || !signAccount.address || status !== TransactionStatusType.SUCCEED) return
+
+                        const result = await connection?.deploy?.(signAccount.address, signAccount.identifier, {
+                            chainId,
+                        })
+
+                        Wallet?.addWallet({
```
Comment (guanbinrui) [inline packages/plugins/SmartPay/src/hooks/useDeploy.tsx]:
```suggestion
                        await Wallet?.addWallet({
```

File: packages/plugins/SmartPay/src/hooks/useDeploy.tsx [hunk 7]
```diff
@@ -101,29 +101,33 @@ export function useDeploy(
...
+                            hasStoredKeyInfo: false,
+                            id: contractAccount.address,
+                            createdAt: new Date(),
+                            updatedAt: new Date(),
+                        }).then(() => {
+                            onSuccess?.()
+                            resolve(result)
+                        })
+                    } catch (error) {
+                        console.log(error)
```
Comment (guanbinrui) [inline packages/plugins/SmartPay/src/hooks/useDeploy.tsx]:
```suggestion
```

File: packages/web3-providers/src/SmartPay/apis/SmartPayAccountAPI.ts [hunk 8]
```diff
@@ -203,29 +203,34 @@ export class SmartPayAccountAPI implements AbstractAccountAPI.Provider<NetworkPl
...
-            await userTransaction.fillTransaction(web3)
-            return this.web3.sendSignedTransaction(chainId, await userTransaction.signTransaction(web3, signer))
+                await userTransaction.fillUserOperation(web3, await getOverrides())
+                return this.bundler.sendUserOperation(chainId, await userTransaction.signUserOperation(signer))
+            } else {
+                await userTransaction.fillTransaction(web3)
+                return this.web3.sendSignedTransaction(chainId, await userTransaction.signTransaction(web3, signer))
+            }
+        } catch (error) {
+            console.log(error)
```
Comment (guanbinrui) [inline packages/web3-providers/src/SmartPay/apis/SmartPayAccountAPI.ts]: Why can't we throw the error?

File: packages/web3-shared/base/src/specs/index.ts [hunk 9]
```diff
@@ -904,7 +904,7 @@ export interface ProviderEvents<ChainId, ProviderType> {
     /** Emit when the chain id changed. */
     chainId: [string]
     /** Emit when the accounts changed. */
-    accounts: [string[]]
+    accounts: [string[], string | void]
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Does it break the consistent with EIP-1193?

File: packages/web3-shared/evm/src/libs/ConnectionContext.ts [hunk 10]
```diff
@@ -107,6 +107,7 @@ export class ConnectionContext {
         return {
             ...this.payloadEditor.config,
             ...this._options?.overrides,
+            from: this._options?.overrides?.from || this.payloadEditor.config?.from,
```
Comment (guanbinrui) [inline packages/web3-shared/evm/src/libs/ConnectionContext.ts:110]:
```suggestion
            from: this._options?.overrides?.from || this.payloadEditor.config?.from,
            chainId: this._options?.overrides?.from || this.payloadEditor.config?.chainId,
```

File: packages/web3-shared/evm/src/libs/UserTransaction.ts [hunk 11]
```diff
@@ -196,7 +196,7 @@ export class UserTransaction {
         }
 
         // 1.5x scale up callGas
-        this.userOperation.callGas = toHex(toFixed(multipliedBy(this.userOperation.callGas ?? '0', 1.5)))
+        this.userOperation.callGas = toHex(multipliedBy(this.userOperation.callGas ?? '0', 1.5).toFixed(0))
```
Comment (guanbinrui) [inline packages/web3-shared/evm/src/libs/UserTransaction.ts]: Please fix `toFixed` by adding some failing cases here:

`packages/web3-shared/base/tests/number.ts`

File: packages/web3-state/src/Provider.ts [hunk 12]
```diff
@@ -102,11 +102,13 @@ export class ProviderState<
                     account,
                 })
             })
-            provider.emitter.on('accounts', async (accounts) => {
+            provider.emitter.on('accounts', async (accounts, chainId) => {
```
Comment (guanbinrui) [inline packages/web3-state/src/Provider.ts]: It breaks the consistent with EIP-1193.

File: packages/plugins/SmartPay/src/hooks/useDeploy.tsx (hunk 4 above)
Comment (nuanyang233) [reply packages/plugins/SmartPay/src/hooks/useDeploy.tsx:45]: Avoid being contaminated by account on the chain context.

File: packages/web3-providers/src/SmartPay/apis/SmartPayAccountAPI.ts (hunk 8 above)
Comment (nuanyang233) [reply packages/web3-providers/src/SmartPay/apis/SmartPayAccountAPI.ts]: Will be remove.

File: packages/plugins/EVM/src/state/Connection/providers/BaseHosted.ts [hunk 13]
```diff
@@ -65,7 +65,11 @@ export class BaseHostedProvider extends BaseProvider implements EVM_Provider {
     }
 
     private onAccountChanged() {
-        if (this.hostedAccount) this.emitter.emit('accounts', [this.hostedAccount])
+        // if (this.hostedAccount) this.emitter.emit('accounts', [this.hostedAccount], toHex(this.hostedChainId))
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Connection/providers/BaseHosted.ts]:
```suggestion
```

File: packages/plugins/EVM/src/state/Wallet.ts [hunk 14]
```diff
@@ -77,6 +78,16 @@ export class Wallet extends WalletState<ProviderType, Transaction> {
                     this.storage.value[ProviderType.MaskWallet]?.find((item) => isSameAddress(item.address, x.address))
                         ?.name ?? x.name,
             }))
+
+            const account = Web3StateSettings.value.Provider?.account?.getCurrentValue()
+
+            // If the currently connected account no longer exists, then disconnect
+            if (result.length && account && !result.find((x) => isSameAddress(x.address, account))) {
+                const connection = Web3StateSettings.value.Connection?.getConnection?.()
+                connection?.disconnect()
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Wallet.ts]:
```suggestion
                await connection?.disconnect()
```

**Pull request merged**